tests/ui/cast.rs

```diff
@@ -61,4 +61,15 @@ fn main() {
     1u32 as usize; // Should not trigger any lint
     1i32 as isize; // Neither should this
     1i32 as usize;
+
+    // Test cast_unnecessary
+    1i32 as i32;
+    1f32 as f32;
+    false as bool;
+    &1i32 as &i32;
+
+    1i32 as i64;      // Should not trigger
```

Comment:
I printed `cx.tables.cast_kinds.get(&ex.id)` for the cast expression.
Unfortunately, looks like both `1i32 as i32` and `1 as u64` are coercion cast, no way to tell the difference.

Need more thoughts.

Here is from the book, maybe helpful:
> The type of an unsuffixed integer literal is determined by type inference:
> If an integer type can be uniquely determined from the surrounding program context, the unsuffixed integer literal has that type.
> If the program context under-constrains the type, it defaults to the signed 32-bit integer i32.
> If the program context over-constrains the type, it is considered a static type error.